consensus/spos/commonSubround/subroundStartRound.go

```diff
@@ -255,3 +270,42 @@ func (sr *SubroundStartRound) generateNextConsensusGroup(roundIndex int64) error
 
 	return nil
 }
+
+// EpochStartPrepare wis called when an epoch start event is observed, but not yet confirmed/committed.
+// Some components may need to do initialisation on this event
+func (sr *SubroundStartRound) EpochStartPrepare(metaHeader data.HeaderHandler) {
+	log.Trace(fmt.Sprintf("epoch %d start prepare in consensus", metaHeader.GetEpoch()))
+}
+
+// EpochStartAction is called upon a start of epoch event.
+func (sr *SubroundStartRound) EpochStartAction(hdr data.HeaderHandler) {
+	log.Trace(fmt.Sprintf("epoch %d start action in consensus", hdr.GetEpoch()))
+
+	sr.changeEpoch(hdr)
+}
+
+func (sr *SubroundStartRound) changeEpoch(header data.HeaderHandler) {
+	publicKeysPrevEpoch, err := sr.NodesCoordinator().GetAllValidatorsPublicKeys(header.GetEpoch() - 1)
+	if err != nil {
+		log.Error(fmt.Sprintf("epoch %d: %s", header.GetEpoch()-1, err.Error()))
+		return
+	}
+
+	publicKeysNewEpoch, err := sr.NodesCoordinator().GetAllValidatorsPublicKeys(header.GetEpoch())
+	if err != nil {
+		log.Error(fmt.Sprintf("epoch %d: %s", header.GetEpoch(), err.Error()))
+		return
+	}
+
+	shardEligible := make(map[string]struct{}, 0)
+	shardId := sr.ShardCoordinator().SelfId()
```

Comment:
`// TODO update this when Shard Shuffling is implemented` :P

Reply:
right, added the todo